src/util/newconfig/ConfigDefinition.hpp

```diff
@@ -0,0 +1,187 @@
+ */
+class ClioConfigDefinition {
+public:
+    using KeyValuePair = std::pair<std::string_view, std::variant<ConfigValue, Array>>;
+
+    /**
+     * @brief Constructs a new ClioConfigDefinition
+     *
+     * Initializes the configuration with a predefined set of key-value pairs
+     * If a key contains "[]", the corresponding value must be an Array
+     *
+     * @param pair A list of key-value pairs for the predefined set of clio configurations
+     */
+    ClioConfigDefinition(std::initializer_list<KeyValuePair> pair)
+    {
+        for (auto const& p : pair) {
+            if (p.first.contains("[]"))
+                ASSERT(std::holds_alternative<Array>(p.second), "value must be array if key has \"[]\"");
+            map_.insert(p);
+        }
+    }
+
+    /**
+     * @brief Parses the configuration file
+     *
+     * Should also check that no extra configuration key/value pairs are present
+     *
+     * @param config The configuration file interface
+     * @return An optional Error object if parsing fails
+     */
+    std::optional<Error>
+    parse(ConfigFileInterface const& config);
+
+    /**
+     * @brief Validates the configuration file
+     *
+     * Should only check for valid values, without populating
+     *
+     * @param config The configuration file interface
+     * @return An optional Error object if validation fails
+     */
+    std::optional<Error>
+    validate(ConfigFileInterface const& config) const;
+
+    /**
+     * @brief Generate markdown file of all the clio config descriptions
+     *
+     * @param configDescription The configuration description object
+     * @return An optional Error if generating markdown fails
+     */
+    std::expected<std::string, Error>
+    getMarkDown(ClioConfigDescription const& configDescription) const;
+
+    /**
+     * @brief Returns the ObjectView specified with the prefix
+     *
+     * @param prefix The key prefix for the ObjectView
+     * @param idx Used if getting Object in an Array
+     * @return ObjectView with the given prefix
+     */
+    ObjectView
+    getObject(std::string_view prefix, std::optional<std::size_t> idx = std::nullopt) const;
+
+    /**
+     * @brief Returns the specified ValueView object associated with the key
+     *
+     * @param fullKey The config key to search for
+     * @return ValueView associated with the given key
+     */
+    ValueView
+    getValue(std::string_view fullKey) const;
+
+    /**
+     * @brief Returns the specified Array object from ClioConfigDefinition
+     *
+     * @param prefix The prefix to search config keys from
+     * @return ArrayView with all key-value pairs where key starts with "prefix"
+     */
+    ArrayView
+    getArray(std::string_view prefix) const;
+
+    /**
+     * @brief Returns an iterator to the beginning of the configuration map.
+     *
+     * @return A constant iterator to the beginning of the map.
+     */
+    auto
+    begin() const
```

Comment:
Are these `begin()` and `end()` used?

Reply:
Used in ObjectView for method "startsWithKey"